File: tests-cypress/cypress/support/pages/markets/markets-page.ts
import { routes } from '../../constants/routes'
import { testId } from '../../utils/core/selectors'

export const marketsObj = {
  page: testId('markets-page'),
  total: testId('markets-total'),
  addButton: testId('coin-add-button'),
  search: testId('markets-search'),
  filterCategory: testId('markets-filter-category'),
  filterStatus: testId('markets-filter-status'),
  pageSize: testId('markets-page-size'),
  error: testId('markets-error'),
  table: testId('markets-table'),
  empty: testId('markets-empty'),
  deleteModal: {
    modal: testId('coin-delete-modal'),
    cancel: testId('coin-delete-cancel'),
    confirm: testId('coin-delete-confirm'),
  },
  pagination: {
    nav: testId('pagination'),
    prev: testId('pagination-prev'),
    next: testId('pagination-next'),
  },
  sortHeader: (column: string) => testId(`markets-sort-${column}`),
  row: (symbol: string) => testId(`markets-row-${symbol}`),
  watchlistToggle: (symbol: string) => testId(`watchlist-toggle-${symbol}`),
  editButton: (symbol: string) => testId(`coin-edit-${symbol}`),
  deleteButton: (symbol: string) => testId(`coin-delete-${symbol}`),
  paginationPage: (oneBasedPage: number) => testId(`pagination-page-${oneBasedPage}`),
}

export const marketsPage = {
  visit: () => {
    cy.visit(routes.markets)
    cy.get(marketsObj.page).should('be.visible')
    cy.get(marketsObj.table).should('be.visible')
  },

  search: (term: string) => {
    cy.get(marketsObj.search).clear().type(term)
  },

  filterByCategory: (category: string) => {
    cy.get(marketsObj.filterCategory).select(category)
  },

  filterByStatus: (status: string) => {
    cy.get(marketsObj.filterStatus).select(status)
  },

  setPageSize: (size: string) => {
    cy.get(marketsObj.pageSize).select(size)
  },

  sortBy: (column: string) => {
    cy.get(marketsObj.sortHeader(column)).click()
  },

  goToNextPage: () => {
    cy.get(marketsObj.pagination.next).click()
  },

  toggleWatchlist: (symbol: string) => {
    cy.get(marketsObj.watchlistToggle(symbol)).click()
  },

  openDeleteModal: (symbol: string) => {
    cy.get(marketsObj.deleteButton(symbol)).click()
    cy.get(marketsObj.deleteModal.modal).should('be.visible')
  },

  cancelDelete: () => {
    cy.get(marketsObj.deleteModal.cancel).click()
    cy.get(marketsObj.deleteModal.modal).should('not.exist')
  },

  confirmDelete: () => {
    cy.get(marketsObj.deleteModal.confirm).click()
    cy.get(marketsObj.deleteModal.modal).should('not.exist')
  },

  verifyShellReady: () => {
    cy.get(marketsObj.table).should('be.visible')
    cy.get(marketsObj.search).should('be.visible')
    cy.get(marketsObj.filterCategory).should('be.visible')
  },

  verifyFirstRow: (symbol: string) => {
    cy.get(`${marketsObj.table} tbody tr`)
      .first()
      .should('have.attr', 'data-testid', `markets-row-${symbol}`)
  },

  verifySortIndicator: (column: string, direction: 'asc' | 'desc') => {
    cy.get(marketsObj.sortHeader(column)).should('contain.text', direction === 'asc' ? '▲' : '▼')
  },

  verifyRowVisible: (symbol: string) => {
    cy.get(marketsObj.row(symbol)).should('be.visible')
  },

  verifyRowAbsent: (symbol: string) => {
    cy.get(marketsObj.row(symbol)).should('not.exist')
  },

  verifyEmptyState: () => {
    cy.get(marketsObj.empty).should('be.visible')
  },

  verifyTotalCount: (count: number) => {
    cy.get(marketsObj.total).should('have.text', `${count} coins`)
  },

  verifyPaginationVisible: () => {
    cy.get(marketsObj.pagination.nav).should('be.visible')
  },

  verifyWatchlistStarOn: (symbol: string) => {
    cy.get(marketsObj.watchlistToggle(symbol)).should('have.class', 'markets__star--on')
  },

  verifyWatchlistStarOff: (symbol: string) => {
    cy.get(marketsObj.watchlistToggle(symbol)).should('not.have.class', 'markets__star--on')
  },

  verifyAddButtonVisible: () => {
    cy.get(marketsObj.addButton).should('be.visible')
  },

  verifyAddButtonAbsent: () => {
    cy.get(marketsObj.addButton).should('not.exist')
  },

  verifyAdminActionsVisible: (symbol: string) => {
    cy.get(marketsObj.editButton(symbol)).should('be.visible')
    cy.get(marketsObj.deleteButton(symbol)).should('be.visible')
  },

  verifyAdminActionsAbsent: (symbol: string) => {
    cy.get(marketsObj.editButton(symbol)).should('not.exist')
    cy.get(marketsObj.deleteButton(symbol)).should('not.exist')
  },
}

File: tests-selenium/src/test/java/com/coinpulse/selenium/pages/MarketsPage.java
package com.coinpulse.selenium.pages;

import com.coinpulse.selenium.constants.Routes;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;

/**
 * Markets page object; selectors mirror the Cypress markets page object.
 */
public class MarketsPage extends BasePage {

    private static final By PAGE = byTestId("markets-page");
    private static final By TABLE = byTestId("markets-table");
    private static final By SEARCH = byTestId("markets-search");
    private static final By FILTER_CATEGORY = byTestId("markets-filter-category");
    private static final By FILTER_STATUS = byTestId("markets-filter-status");
    private static final By EMPTY = byTestId("markets-empty");
    private static final By ADD_BUTTON = byTestId("coin-add-button");
    private static final By DELETE_MODAL = byTestId("coin-delete-modal");
    private static final By DELETE_CANCEL = byTestId("coin-delete-cancel");
    private static final By DELETE_CONFIRM = byTestId("coin-delete-confirm");
    private static final By FIRST_ROW = By.cssSelector("[data-testid='markets-table'] tbody tr");

    private static final String ROW_TESTID_PREFIX = "markets-row-";
    private static final String STAR_ON_CLASS = "markets__star--on";
    private static final String SORT_ASC_INDICATOR = "▲";
    private static final String SORT_DESC_INDICATOR = "▼";

    public MarketsPage(WebDriver driver) {
        super(driver);
    }

    private static By row(String symbol) {
        return byTestId(ROW_TESTID_PREFIX + symbol);
    }

    private static By sortHeader(String column) {
        return byTestId("markets-sort-" + column);
    }

    private static By watchlistToggle(String symbol) {
        return byTestId("watchlist-toggle-" + symbol);
    }

    private static By editButton(String symbol) {
        return byTestId("coin-edit-" + symbol);
    }

    private static By deleteButton(String symbol) {
        return byTestId("coin-delete-" + symbol);
    }

    public void open() {
        openPath(Routes.MARKETS);
        awaitVisible(PAGE);
        awaitVisible(TABLE);
    }

    public void awaitLoaded() {
        awaitVisible(PAGE);
        awaitVisible(TABLE);
        awaitPath(Routes.MARKETS);
    }

    public void verifyShellReady() {
        awaitVisible(TABLE);
        awaitVisible(SEARCH);
        awaitVisible(FILTER_CATEGORY);
    }

    public void search(String term) {
        type(SEARCH, term);
    }

    public void filterByStatus(String status) {
        selectByValue(FILTER_STATUS, status);
    }

    public void sortBy(String column) {
        click(sortHeader(column));
    }

    public void toggleWatchlist(String symbol) {
        click(watchlistToggle(symbol));
    }

    public void openDeleteModal(String symbol) {
        click(deleteButton(symbol));
        awaitVisible(DELETE_MODAL);
    }

    public void cancelDelete() {
        click(DELETE_CANCEL);
        awaitAbsent(DELETE_MODAL);
    }

    public void confirmDelete() {
        click(DELETE_CONFIRM);
        awaitAbsent(DELETE_MODAL);
    }

    public void verifyFirstRow(String symbol) {
        wait.until(ExpectedConditions.attributeToBe(FIRST_ROW, "data-testid",
                ROW_TESTID_PREFIX + symbol));
    }

    public void verifySortIndicator(String column, String direction) {
        awaitTextContains(sortHeader(column),
                "asc".equals(direction) ? SORT_ASC_INDICATOR : SORT_DESC_INDICATOR);
    }

    public void verifyRowVisible(String symbol) {
        awaitVisible(row(symbol));
    }

    public void verifyRowAbsent(String symbol) {
        awaitAbsent(row(symbol));
    }

    public void verifyEmptyState() {
        awaitVisible(EMPTY);
    }

    public void verifyWatchlistStarOn(String symbol) {
        wait.until(d -> d.findElement(watchlistToggle(symbol))
                .getAttribute("class").contains(STAR_ON_CLASS));
    }

    public void verifyWatchlistStarOff(String symbol) {
        wait.until(d -> !d.findElement(watchlistToggle(symbol))
                .getAttribute("class").contains(STAR_ON_CLASS));
    }

    public void verifyAddButtonVisible() {
        awaitVisible(ADD_BUTTON);
    }

    public void verifyAddButtonAbsent() {
        awaitAbsent(ADD_BUTTON);
    }

    public void verifyAdminActionsVisible(String symbol) {
        awaitVisible(editButton(symbol));
        awaitVisible(deleteButton(symbol));
    }

    public void verifyAdminActionsAbsent(String symbol) {
        awaitVisible(row(symbol));
        awaitAbsent(editButton(symbol));
        awaitAbsent(deleteButton(symbol));
    }
}
